java/ql/src/experimental/CWE-939/IncorrectURLVerification.ql

```diff
@@ -0,0 +1,87 @@
+/**
+ * @id java/incorrect-url-verification
+ * @name Incorrect URL verification
+ * @description Apps that rely on URL parsing to verify that a given URL is pointing to a trusted server are susceptible to wrong ways of URL parsing and verification.
+ * @kind problem
+ * @tags security
+ *       external/cwe-939
+ */
+
+import java
+
+/**
+ * The Java class `android.net.Uri` and `java.net.URL`.
+ */
+class Uri extends RefType {
+  Uri() {
+    hasQualifiedName("android.net", "Uri") or
+    hasQualifiedName("java.net", "URL")
+  }
+}
+
+/**
+ * The method `getHost()` declared in `android.net.Uri` and `java.net.URL`.
+ */
+class UriGetHostMethod extends Method {
+  UriGetHostMethod() {
+    getDeclaringType() instanceof Uri and
+    hasName("getHost") and
+    getNumberOfParameters() = 0
+  }
+}
+
+/**
+ * The method access with incorrect string comparision
+ */
+class HostVerificationMethodAccess extends MethodAccess {
+  HostVerificationMethodAccess() {
+    (
+      this.getMethod().hasName("endsWith") or
+      this.getMethod().hasName("contains") or
+      this.getMethod().hasName("indexOf")
+    ) and
+    this.getMethod().getNumberOfParameters() = 1 and
+    (
+      this.getArgument(0).(StringLiteral).getRepresentedString().charAt(0) != "." //string constant comparison e.g. uri.getHost().endsWith("example.com")
+      or
+      this
+          .getArgument(0)
+          .(AddExpr)
+          .getLeftOperand()
+          .(VarAccess)
+          .getVariable()
+          .getAnAssignedValue()
+          .(StringLiteral)
+          .getRepresentedString()
+          .charAt(0) != "." //var1+var2, check var1 starts with "." e.g. String domainName = "example"; Uri.parse(url).getHost().endsWith(domainName+".com")
+      or
+      this
+          .getArgument(0)
+          .(AddExpr)
+          .getLeftOperand()
+          .(StringLiteral)
+          .getRepresentedString()
+          .charAt(0) != "." //"."+var2, check string constant "." e.g. String domainName = "example.com";  Uri.parse(url).getHost().endsWith("www."+domainName)
+      or
+      exists(MethodAccess ma |
+        this.getArgument(0) = ma and
+        ma.getMethod().hasName("getString") and
+        ma.getArgument(0).toString().indexOf("R.string") = 0
+      ) //Check resource properties in /res/values/strings.xml in Android mobile applications using res.getString(R.string.key)
```

Comment:
I still don't know what `R.string` is - maybe it's just me who don't know about Android programming.  In any case, matching using `.toString()` here is not good practise and similarly matching a method only by name its name isn't either.  For both cases you should include the fully qualified declaring types of the method and fields you intend to match.  That way it will be possible to at least google to find information on the methods and fields being matched.  E.g.
```
exists(..., Method m, Field f |
...
ma.getMethod() = m and m.hasName("getString") and m.getDeclaringType().hasQualifiedName(....)
...
ma.getArgument(0).(FieldRead).getQualifier().(FieldRead).getField() = f and
f.hasName(...) and f.getDeclaringType().hasQualifiedName(...)
```

Reply:
Hi @aschackmull ,

Thanks for reviewing the query. I've added the declaring type "android.content.res.Resources" to the Android getString() search, which helps to clarify the query, but I'm not sure how Field can help.

For Android development, all projects use a standard [XML structure](https://developer.android.com/training/basics/supporting-devices/languages) to keep localized and formatted strings, e.g.
+ MyProject/
+    res/
+        values/
+            strings.xml
+        values-de/
+            strings.xml
+        values-fr/
+            strings.xml
It's a common practice to configure system properties in this XML file just like using properties files in standalone and web applications.

One example having a trusted server domain configured in strings.xml is [jbutz/AndroidWebViewDemo](https://github.com/jbutz/AndroidWebViewDemo), which has:
```
<resources>
    <string name="app_name">JasonButz WebApp Demo</string>
    <string name="url">file:///android_asset/index.html</string>
    <string name="host">jasonbutz.info</string>
</resources>
```
Resources can be referenced in the source code using each resource's name attribute, e.g. getResources().getString(R.string.host).

Then in its [Java program](https://github.com/jbutz/AndroidWebViewDemo/blob/e2b7d418d61dd26f84a4603943f0e45d371f61e2/WebAppDemo/src/info/jasonbutz/android/web/Main.java), it compares FQDN of an incoming request with the configured domain name:
```
import android.content.res.Resources;
...
Resources res = getResources();
if (Uri.parse(url).getHost().contains(res.getString(R.string.host)) ||
    Uri.parse(url).getScheme().contains("file")) {
    // This is my web site, so do not override; let my WebView
    // load the page
    return false;
}
```

So this query segment is intended to detect such a scenario. It seems that the file strings.xml is not included in the CodeQL database and the compilation time value of R.string.host is not available either. Or else, a more accurate query would be to get the real argument value of res.getString() then check whether it starts with a dot rather than firing the alarm if it's configured in the resources file.

I can delete this query segment if it's a better idea to include it at a later time when the more accurate query can be done. Or is there another way to optimize it?

Thanks,
@luchua-bc

Reply:
I'm just trying to make it clear what is being matched, and this entails that we shouldn't rely on `.toString()`.  I've found https://developer.android.com/reference/android/R.string, which helps me understand a bit more what the `"R.string"` is that you're matching in your example `res.getString(R.string.key)`.  Now it looks to me like what you want is something like
```
ma.getArgument(0).(FieldRead).getField() = f and
f.getDeclaringType() instanceof AndroidRString
...
class AndroidRString extends RefType {
  AndroidRString() { this.hasQualifiedName("android", "R$string") }
}
```

Reply:
Thank you very much for the code of AndroidRString search. I've modified the query as discussed below.

Cheers,
@luchua-bc

Reply:
The `R.string` reference is much clearer now, thanks.